freqtrade/freqai/data_drawer.py

```diff
@@ -78,25 +79,53 @@ def __init__(self, full_path: Path, config: Config, follow_mode: bool = False):
         self.historic_predictions_bkp_path = Path(
             self.full_path / "historic_predictions.backup.pkl")
         self.pair_dictionary_path = Path(self.full_path / "pair_dictionary.json")
+        self.metric_tracker_path = Path(self.full_path / "metric_tracker.json")
         self.follow_mode = follow_mode
         if follow_mode:
             self.create_follower_dict()
         self.load_drawer_from_disk()
         self.load_historic_predictions_from_disk()
+        self.load_metric_tracker_from_disk()
         self.training_queue: Dict[str, int] = {}
         self.history_lock = threading.Lock()
         self.save_lock = threading.Lock()
         self.pair_dict_lock = threading.Lock()
+        self.metric_tracker_lock = threading.Lock()
         self.old_DBSCAN_eps: Dict[str, float] = {}
         self.empty_pair_dict: pair_info = {
                 "model_filename": "", "trained_timestamp": 0,
                 "data_path": "", "extras": {}}
+        self.metric_tracker: Dict[str, Dict[str, list]] = {}
+
+    def update_metric_tracker(self, metric: str, value: float, pair: str) -> None:
+        """
+        General utility for adding and updating custom metrics. Typically used
+        for adding training performance, train timings, inferenc timings, cpu loads etc.
+        """
+        with self.metric_tracker_lock:
+            if pair not in self.metric_tracker:
+                self.metric_tracker[pair] = {}
+            if metric not in self.metric_tracker[pair]:
+                self.metric_tracker[pair][metric] = []
+
+            self.metric_tracker[pair][metric].append(value)
+
+    def collect_metrics(self, time_spent: float, pair: str):
+        """
+        Add metrics to the metric tracker dictionary
+        """
+        load1, load5, load15 = psutil.getloadavg()
+        cpus = psutil.cpu_count()
+        self.update_metric_tracker('train_time', time_spent, pair)
```

Comment:
how often do we assume this is called at max? (per second / minute / hour / ...).

we have to treat freqtrade as a long-running process - which could be running for days/weeks/months without a restart.
this (or rather, update_metric_tracker) has every sign of a potential memory leak.

while the information in it seems negligible at first glance - i'm not sure it is if this is called frequently enough. 
we save a 4 floats (8bytes) per call. Assuming this is called once per second, then the object-structure will grow by `2_764_800` bytes (~2 mb) per day/pair - on a datastructure that's kept in memory - and loaded from disk on restart.

Reply:
This is called once per train per coin and once per pair list inference (once per base candle). So the amount depends on train frequency and base timeframe.

Reply:
so assuming we're on the 1m timeframe, we're down to ~46kb/day/pair - which i guess is ok, especially considering this needs to be explicitly activated.